python/test/test_binding.py

```diff
@@ -0,0 +1,1152 @@
+
+        t_list = [
+            (
+                triton_bindings.TRITONSERVER_InferenceTraceActivity.
+                REQUEST_START,
+                "REQUEST_START",
+            ),
+            (
+                triton_bindings.TRITONSERVER_InferenceTraceActivity.QUEUE_START,
+                "QUEUE_START",
+            ),
+            (
+                triton_bindings.TRITONSERVER_InferenceTraceActivity.
+                COMPUTE_START,
+                "COMPUTE_START",
+            ),
+            (
+                triton_bindings.TRITONSERVER_InferenceTraceActivity.
+                COMPUTE_INPUT_END,
+                "COMPUTE_INPUT_END",
+            ),
+            (
+                triton_bindings.TRITONSERVER_InferenceTraceActivity.
+                COMPUTE_OUTPUT_START,
+                "COMPUTE_OUTPUT_START",
+            ),
+            (
+                triton_bindings.TRITONSERVER_InferenceTraceActivity.COMPUTE_END,
+                "COMPUTE_END",
+            ),
+            (
+                triton_bindings.TRITONSERVER_InferenceTraceActivity.REQUEST_END,
+                "REQUEST_END",
+            ),
+            (
+                triton_bindings.TRITONSERVER_InferenceTraceActivity.
+                TENSOR_QUEUE_INPUT,
+                "TENSOR_QUEUE_INPUT",
+            ),
+            (
+                triton_bindings.TRITONSERVER_InferenceTraceActivity.
+                TENSOR_BACKEND_INPUT,
+                "TENSOR_BACKEND_INPUT",
+            ),
+            (
+                triton_bindings.TRITONSERVER_InferenceTraceActivity.
+                TENSOR_BACKEND_OUTPUT,
+                "TENSOR_BACKEND_OUTPUT",
+            ),
+        ]
+        for t, t_str in t_list:
+            self.assertEqual(
+                triton_bindings.TRITONSERVER_InferenceTraceActivityString(t),
+                t_str)
+
+    def test_trace(self):
+        # This test depends on 'test_infer_async' test to capture
+        # the trace
+        level = int(
+            triton_bindings.TRITONSERVER_InferenceTraceLevel.TIMESTAMPS) | int(
+                triton_bindings.TRITONSERVER_InferenceTraceLevel.TENSORS)
+        trace_dict = {"signal_queue": queue.Queue()}
+        trace = triton_bindings.TRITONSERVER_InferenceTrace(
+            level, 123, g_timestamp_fn, g_tensor_fn, g_trace_release_fn,
+            trace_dict)
+        # [FIXME] get a copy of trace id due to potential issue of 'trace'
+        # lifecycle
+        trace_id = trace.id
+
+        # Send and wait for inference, not care about result.
+        server = self._start_polling_server()
+        (
+            request,
+            allocator,
+            response_queue,
+            request_counter,
+        ) = self._prepare_inference_request(server)
+        server.infer_async(request, trace)
+
+        # [FIXME] WAR due to trace lifecycle is tied to response in Triton core,
+        # trace reference should drop on response send..
```

Comment:
Yes, the coupling is within core logic and we need to make change there to decouple the lifecycle. This is not obvious in direct C API usage because users typically delete response on response complete callback which triggers trace release callback, in Python the time of deletion is less-deterministic and can cause confusion when trace release is being expected.